File: solutions/day01/day1_solution.py
class Solution:
    def numWaterBottles(self, numBottles, numExchange):
        """
        Calculates the total number of water bottles you can drink given an initial 
        number of full bottles and a rule for exchanging empty bottles for new full bottles.

        Parameters:
        numBottles (int): Initial number of full water bottles.
        numExchange (int): Number of empty bottles required to exchange for one full bottle.

        Returns:
        int: Total number of water bottles you can drink.
        """

        count = 0          # Total bottles drank so far
        emptyBottles = 0   # Empty bottles we currently have

        while True:
            # Drink all current full bottles
            emptyBottles += numBottles
            count += numBottles

            # Exchange empty bottles for new full bottles
            # divmod returns (quotient, remainder)
            newExchanged, leftovers = divmod(emptyBottles, numExchange)

            # If no new bottles can be exchanged, we are done
            if newExchanged == 0:
                break

            # Update empty bottles for the next iteration
            emptyBottles = leftovers
            # Update number of full bottles to drink next round
            numBottles = newExchanged

        return count

File: solutions/day01/Day1_Solution.java
class Solution {
public int numWaterBottles(int numBottles, int numExchange) {

   //drinkedBottles store the total number of bottles you can drink
   int drinkedBottles=0;

   //emptyBottles store the number of empty bottles you currently have
   int emptyBottles=0;

    while(numBottles!=0)
    {
        // Drink all the current full bottles you have
        drinkedBottles+=numBottles;

        // All the bottles you drank now become empty bottles
        emptyBottles+=numBottles;

        // Exchange empty bottles for new full bottles
        numBottles=emptyBottles/numExchange;

        // Update the number of empty bottles left after exchange
        emptyBottles=emptyBottles%numExchange;
        
    }

     // Return the total number of bottles you were able to drink
    return drinkedBottles;

        
    }
}